Allow ingest and editing of line-level annotations (#1504)

blms opened the pull request:
## In this PR

Per #1504:
- Parse the line geometry and metadata in eScriptorium output ALTO XML to create individual annotations per line
- Use the `textGranularity` property to indicate block- or line-level annotation granularity, and a new foreign-key relation between Annotations to represent relationships between the two
  - Update HTML serializer to iterate over only block-level annotations, and if they have associated line-level annotations, compile those lines' contents into an ordered list
- Pass a “line mode” parameter to Tahqiq to tell it when to switch to line-level functionality

see also https://github.com/Princeton-CDH/annotorious-tahqiq/pull/31

## Notes and questions

- Is `partOf` a good field name for the associated block's ID in a line-level annotation? It’s not standard W3C, as that field name typically refers to an AnnotationCollection that an AnnotationPage is part of, but not sure how else to indicate in the serializer what the associated block should be for a line.
- I tried to keep it flexible enough that `textGranularity` could remain null for normal (block-level) annotations, to avoid having to update code everywhere annotations are created/parsed/etc. Also avoids having to write a data migration to add block granularity to all existing annotations.

rlskoeser reviewed (approved):
Congratulations on all this progress.

I like `partOf` for the relationship between lines and the blocks they belong to. I never got an answer for how other people are modeling this kind of thing and couldn't find any examples.  If and when we ever share our annotations for other systems, we can determine at that point if there's a more standard format we can use for the public versions.  Can you find somewhere to document this, though? Maybe in the `Annotation` model docstring? (... come to think of it, we should probably have clearer docs on the other things we're doing, like the ordering, and I'm not sure how well it is documented)

I also like your solution for making paragraph/block level granularity the default. Should this assumption be documented somewhere on `Annotation`?

Let me know if there's anything you want me to review in more detail or test locally.

blms commented on geniza/annotations/models.py:
```diff
@@ -79,6 +79,13 @@ class Annotation(TrackChangesModel):
+        to="Annotation",
+        on_delete=models.CASCADE,
+        related_name="lines",
+        null=True,
```
The case where `block` is null always indicates that it's a block level annotation, at least based on the current ingest code. In fact, maybe that's a simpler condition for the queryset than checking `content__textGranularity`…

blms commented on geniza/corpus/management/commands/escr_alto_to_annotation.py:
```diff
@@ -232,33 +272,32 @@ def scale_polygon(self, polygon, scale):
+                    "label": block_type,
+                }
+            ]
+            if block_type in self.rotation_tags:
+                # add rotation tag as a CSS class to this block
+                anno_content["target"]["styleClass"] = block_type
```
Not using it for anything yet, but figured I would include it in the ingest code so we don't have to make modifications later to store this info. Eventually it will be used for showing the rotated lines in the line-by-line editor.

blms commented on sitemedia/js/controllers/annotation_controller.js:
```diff
@@ -106,6 +106,7 @@ export default class extends Controller {
             target: canvasURL,
             manifest: config.manifest_base_url + manifestId,
             csrf_token: config.csrf_token,
+            lineMode: config.line_mode,
```
Exactly right. At the moment, it's only possible to have only blocks, or blocks and lines; not possible to have only lines (without blocks grouping them). In the blocks and lines case, it is only possible to have blocks that have no textual body and are only present in order to group lines and add a label to the group.

blms commented on geniza/corpus/views.py:
```diff
@@ -862,6 +862,8 @@ def get_context_data(self, **kwargs):
                     else "translating",
                     "text_direction": text_direction,
+                    # line-by-line mode for eScriptorium sourced transcriptions
+                    "line_mode": "model" in source.source_type.type,
```
Good question. My guess is that we probably won't have the capacity to produce new line-level transcriptions/translations from within PGP, at least not anytime soon, due to the complexity of implementation and the team's current priorities. But I could see importing them from elsewhere that there might be structured transcriptions w/ lines and blocks. In that case, we'd definitely want to make the modification; maybe something like a computed property on the footnote that indicates what kind of annotation content it relates to.

blms commented on geniza/corpus/management/commands/escr_alto_to_annotation.py:
```diff
@@ -128,34 +145,57 @@ def ingest_xml(self, xmlfile):
+                        ),
+                        block=block,
+                        footnote=footnote,
+                    )
```
Basically I just thought the log entries will get quite cluttered if we include the creation of each line as a separate log entry. But maybe that's fine? Nothing else from activity on a line annotation produces log entries on the associated block, so it's probably better to be consistent here…

rlskoeser replied:
I agree it's a lot, but I think consistency is better. If important down the line, we could consider about aggregating related log entries, like you implemented for the document history view.

blms commented on geniza/corpus/views.py:
```diff
@@ -862,6 +862,8 @@ def get_context_data(self, **kwargs):
+                    # line-by-line mode for eScriptorium sourced transcriptions
+                    "line_mode": "model" in source.source_type.type,
                 },
                 # TODO: Add Footnote notes to the following display, if present
```
Still has not been dealt with! (At the moment this doesn't actually do any footnote lookups, only source—I think it's possible in this view to be creating a new transcription/translation without a footnote.)

blms commented on geniza/footnotes/models.py:
```diff
@@ -612,12 +612,24 @@ def content_html(self):
+        for a in self.annotation_set.exclude(
+            # only iterate through block-level annotations; we will group their lines together
+            # if they have lines. (isnull check required to include missing attribute)
+            Q(content__textGranularity__isnull=False)
+            & Q(content__textGranularity="line")
```
If we end up needing this query elsewhere, I would say yes definitely! Probably fine as is for now though. Improved the comment about the isnull check, thanks for drawing attention to it as it's not worded well.

rlskoeser commented on geniza/annotations/models.py:
```diff
@@ -59,15 +59,29 @@ def group_by_manifest(self):
-    """Annotation model for storing annotations in the database."""
+    """Annotation model for storing annotations in the database.
+
+    Annotations may be either block-level or line-level. Block-level annotation is the default;
+    in most cases, a block-level annotation's content is stored as a TextualBody in its `content`
+    JSON.
+
+    However, block-level annotations may also be used to group line-level annotations, in which case
+    they have no textual content themselves, except for an optional label. Instead, their content
+    is serialized by joining TextualBody values from all associated line-level annotations.
+
+    Line-level annotations are associated with blocks via the `block` property, and that relationship
+    is serialized as `partOf` at the root of the line-level annotation."""
```
great documentation here

rlskoeser commented on geniza/annotations/models.py:
```diff
@@ -59,15 +59,29 @@ def group_by_manifest(self):
     #: date last modified
     modified = models.DateTimeField(auto_now=True)
-    #: json content of the annotation
+    #: json content of the annotation. in addition to W3C Web Annotation Data Model properties,
+    #: may also include: "schema:position", which tracks the order of the annotation with respect
+    #: to others in the same block or canvas; and "textGranularity", which indicates whether this
+    #: is a block- or line-level annotation
```
nice, thanks for documenting explicitly

rlskoeser reviewed (approved):
Revisions all look good to me. Thanks for the expanded documentation, super clear.

I forget if I asked on the first review, have you tested how the annotation export works with the line-level content? It will need to be addressed in some way (at least to make sure the export doesn't break) before you import any line-level content into production. (A temporary solution could be skipping line-level transcriptions for export, but I'm wary of that because I've seen solutions that were meant to be temporary solutions last for much longer than hoped.)

blms commented:
Good call @rlskoeser, the exports work fine as they just call the same methods `content_html` and `content_text` that we use in the public site. However, the filenames were showing up as `unknown-author`, so I went ahead and fixed that and added a test for it. Merging as soon as checks pass—thanks for the review!
